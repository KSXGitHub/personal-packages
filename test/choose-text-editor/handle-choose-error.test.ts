import MockedLogger from './lib/mocked-logger'

import {
  IndeterminableTTY,
  NotFound,
  PrefixesParsingFailure,
  InvalidPrefixes,
  NoEditor,
  ChooseError,
  Status,
  handleChooseError
} from '@khai96x/choose-text-editor'

async function setup (error: ChooseError) {
  const { log, getLogs, getText } = new MockedLogger()
  const status = await handleChooseError(log, error)
  return { error, log, getLogs, getText, status }
}

describe('IndeterminableTTY', () => {
  const error = IndeterminableTTY()

  it('calls logError', async () => {
    const { getLogs } = await setup(error)
    expect(getLogs()).toMatchSnapshot()
  })

  it('prints error messages', async () => {
    const { getText } = await setup(error)
    expect(getText()).toMatchSnapshot()
  })

  it('returns status code of IndeterminableTTY', async () => {
    const { status } = await setup(error)
    expect(Status[status]).toBe(Status[Status.IndeterminableTTY])
  })
})

describe('NotFound', () => {
  const error = NotFound()

  it('calls logError', async () => {
    const { getLogs } = await setup(error)
    expect(getLogs()).toMatchSnapshot()
  })

  it('prints error messages', async () => {
    const { getText } = await setup(error)
    expect(getText()).toMatchSnapshot()
  })

  it('returns status code of NotFound', async () => {
    const { status } = await setup(error)
    expect(Status[status]).toBe(Status[Status.NotFound])
  })
})

describe('NoEditor', () => {
  const error = NoEditor()

  it('calls logError', async () => {
    const { getLogs } = await setup(error)
    expect(getLogs()).toMatchSnapshot()
  })

  it('prints error messages', async () => {
    const { getText } = await setup(error)
    expect(getText()).toMatchSnapshot()
  })

  it('returns status code of EmptyEditorSet', async () => {
    const { status } = await setup(error)
    expect(Status[status]).toBe(Status[Status.EmptyEditorSet])
  })
})

describe('PrefixesParsingFailure', () => {
  const error = PrefixesParsingFailure(
    new Error('{error message}'),
    '{invalid prefixes}',
    'FORCE_EDITOR_PREFIXES'
  )

  it('calls logError', async () => {
    const { getLogs } = await setup(error)
    expect(getLogs()).toMatchSnapshot()
  })

  it('prints error messages', async () => {
    const { getText } = await setup(error)
    expect(getText()).toMatchSnapshot()
  })

  it('returns status code of InvalidPrefix', async () => {
    const { status } = await setup(error)
    expect(Status[status]).toBe(Status[Status.InvalidPrefix])
  })
})

describe('InvalidPrefixes', () => {
  async function makeError () {
    const { schemas } = await import('@khai96x/choose-text-editor')
    const invalidCliArguments = ['abc', 123, true, { object: 'illegal' }, ['illegal', 'array']]
    const validatorResult = schemas.CliArguments().validate(invalidCliArguments, {
      allowUnknownAttributes: true
    })
    return InvalidPrefixes(
      validatorResult,
      invalidCliArguments,
      'FORCE_EDITOR_PREFIXES'
    )
  }

  it('calls logError', async () => {
    const { getLogs } = await setup(await makeError())
    expect(getLogs()).toMatchSnapshot()
  })

  it('prints error messages', async () => {
    const { getText } = await setup(await makeError())
    expect(getText()).toMatchSnapshot()
  })

  it('returns status code of InvalidPrefix', async () => {
    const { status } = await setup(await makeError())
    expect(Status[status]).toBe(Status[Status.InvalidPrefix])
  })
})
