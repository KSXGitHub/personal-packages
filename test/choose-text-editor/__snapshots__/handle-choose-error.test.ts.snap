// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IndeterminableTTY calls logError 1`] = `
Array [
  Array [
    "[ERROR] Cannot determine whether terminal is graphical or not",
  ],
  Array [
    "help: You may set ISINTTY=true to use terminal editors, or ISINTTY=false to use graphical editors",
  ],
]
`;

exports[`IndeterminableTTY prints error messages 1`] = `
"
[ERROR] Cannot determine whether terminal is graphical or not
help: You may set ISINTTY=true to use terminal editors, or ISINTTY=false to use graphical editors
"
`;

exports[`InvalidPrefixes calls logError 1`] = `
Array [
  Array [
    "[ERROR] Prefixes does not satisfy its schema",
  ],
  Array [
    "help: Instance must be an array of strings",
  ],
  Array [
    "* env key: 'FORCE_EDITOR_PREFIXES'",
  ],
  Array [
    "* instance: [ 'abc', 123, true, { object: 'illegal' }, [ 'illegal', 'array' ] ]",
  ],
  Array [
    "* stack:",
  ],
  Array [
    "    0: instance[3] is not of a type(s) null,string,number,boolean",
  ],
  Array [
    "    1: instance[4] is not of a type(s) null,string,number,boolean",
  ],
  Array [
    "    ",
  ],
]
`;

exports[`InvalidPrefixes prints error messages 1`] = `
"
[ERROR] Prefixes does not satisfy its schema
help: Instance must be an array of strings
* env key: 'FORCE_EDITOR_PREFIXES'
* instance: [ 'abc', 123, true, { object: 'illegal' }, [ 'illegal', 'array' ] ]
* stack:
    0: instance[3] is not of a type(s) null,string,number,boolean
    1: instance[4] is not of a type(s) null,string,number,boolean
    
"
`;

exports[`NoEditor calls logError 1`] = `
Array [
  Array [
    "[ERROR] No suitable editor",
  ],
  Array [
    "help: When ISINTTY=true, \\"terminal\\" property of your config must not be empty",
  ],
  Array [
    "help: When ISINTTY=false, either \\"graphical\\" or \\"terminal\\" property of your config must not be empty",
  ],
]
`;

exports[`NoEditor prints error messages 1`] = `
"
[ERROR] No suitable editor
help: When ISINTTY=true, \\"terminal\\" property of your config must not be empty
help: When ISINTTY=false, either \\"graphical\\" or \\"terminal\\" property of your config must not be empty
"
`;

exports[`NotFound calls logError 1`] = `
Array [
  Array [
    "[ERROR] No editor detected",
  ],
  Array [
    "help: Check if (at least one of) your editors are installed",
  ],
  Array [
    "help: Check if there is any typo in your config",
  ],
]
`;

exports[`NotFound prints error messages 1`] = `
"
[ERROR] No editor detected
help: Check if (at least one of) your editors are installed
help: Check if there is any typo in your config
"
`;

exports[`PrefixesParsingFailure calls logError 1`] = `
Array [
  Array [
    "[ERROR] Failed to parse prefixes",
  ],
  Array [
    "help: Content must be a valid yaml array of strings",
  ],
  Array [
    "* env key: 'FORCE_EDITOR_PREFIXES'",
  ],
  Array [
    "* env value: '{invalid prefixes}'",
  ],
  Array [
    "* error: Error: {error message}",
  ],
]
`;

exports[`PrefixesParsingFailure prints error messages 1`] = `
"
[ERROR] Failed to parse prefixes
help: Content must be a valid yaml array of strings
* env key: 'FORCE_EDITOR_PREFIXES'
* env value: '{invalid prefixes}'
* error: Error: {error message}
"
`;
